sapp/ui/trace.py

```diff
@@ -403,3 +438,94 @@ def trace_kind_to_shared_text_kind(trace_kind: Optional[TraceKind]) -> SharedTex
+) -> SARIFCodeflowsObject:
+    postcondition_initial_frames = initial_frames(
+        session,
+        issue.issue_instance_id,
+        TraceKind.POSTCONDITION,
+    )
+    precondition_initial_frames = initial_frames(
+        session,
+        issue.issue_instance_id,
+        TraceKind.PRECONDITION,
+    )
+    postcondition_navigation = navigate_trace_frames(
+        session,
+        postcondition_initial_frames,
+        set(issue.source_kinds),
+        set(issue.sink_kinds),
+    )
+    precondition_navigation = navigate_trace_frames(
+        session,
+        precondition_initial_frames,
+        set(issue.source_kinds),
+        set(issue.sink_kinds),
+    )
+    trace_tuples = _create_trace_tuples(
+        reversed(postcondition_navigation)
+    ) + _create_trace_tuples(precondition_navigation)
+    codeflows = [{"threadFlows": [{"locations": []}]}]
+    nesting_level = 0
+    for t in trace_tuples:
+        location = _sarif_codeflow_location_from_trace_tuple(
+            t.trace_frame, tool, nesting_level, True
+        )
+        codeflows[0]["threadFlows"][0]["locations"].append(location)
+        nesting_level += 1
+
+    return codeflows
+
+
+def _create_trace_tuples(
+    navigation: Iterable[Tuple[TraceFrameQueryResult, int]]
+) -> List[TraceTuple]:
+    return [
+        TraceTuple(
+            trace_frame=trace_frame,
+            branches=branches,
+            missing=trace_frame.caller == "",
+        )
+        for trace_frame, branches in navigation
+    ]
+
+
+def _sarif_codeflow_location_from_trace_tuple(
+    trace_frame: TraceFrameQueryResult,
+    tool: str,
+    nesting_level: int = 1,
+    output_features: bool = False,
+) -> Dict[str, Union[SARIFCodeflowLocationObject, Dict[str, str]]]:
+    features_str = titos = ""
+    if output_features:
+        frame_features = [
+            text.contents
+            for text in trace_frame.shared_texts
+            if text.kind is SharedTextKind.FEATURE
+        ]
+        if frame_features:
+            features_str = f"features: {frame_features}"
+        if trace_frame.titos and len(trace_frame.titos.split(";")) > 0:
+            titos = f"via {len(trace_frame.titos.split(';'))} propagators"
+    trace_region = {}
+    if trace_frame.callee_location:
+        trace_region = trace_frame.callee_location.to_sarif()
+    location = {
+        "location": {
+            "physicalLocation": {
+                "artifactLocation": {
+                    "uri": trace_frame.filename,
+                    "uriBaseId": "%SRCROOT%",
+                },
+                "region": trace_region,
+            },
+            "message": {
+                "text": f"flow from {trace_frame.get_clean_caller(tool)}(...{trace_frame.caller_port}...) -[into]-> {trace_frame.get_clean_callee(tool)}(...{trace_frame.callee_port}...) {titos} {features_str}".strip()
+            },
+        },
+        "nestingLevel": nesting_level,
+    }
+    return location
```

Comment:
Could we move all the logic regarding sarif into `sarif.py`?

Reply:
Well IMO anything related to sarif should go into `sarif.py`. If we start putting function from every format into `trace.py` it will get overwhelming soon.
`trace.py` should provide a clean API that allows any formatter to get the data and do it's formatting.
Now I understand this is a pre-existing problem, so we can probably try to fix that later.